src/humanloop/evals/run.py

```diff
@@ -315,61 +219,82 @@ def upload_callback(log_id: str):
-                error_message = str(e).replace("\n", " ")
-                sys.stderr.write(
-                    f"\n{RED}Your {hl_file.type}'s `callable` failed for Datapoint: {dp.id}. Error: {error_message[:100]}...{RESET}\n"
                 )
+                start_time = datetime.now()
+                evaluation_context = get_evaluation_context()
+                if evaluation_context is None:
+                    raise HumanloopRuntimeError(
+                        "Internal error: evaluation context is not set while processing a datapoint."
+                    )
+                try:
+                    output = _call_function(function_, hl_file.type, dp)  # type: ignore [arg-type]
+                    if not evaluation_context.logged:
+                        # function_ did not Log against the source_datapoint_id/ run_id pair
+                        # so we need to create a Log
+                        log = log_func(
+                            **{
+                                "inputs": dp.inputs,
+                                "output": output,
+                                "start_time": start_time,
+                                "end_time": datetime.now(),
+                                "source_datapoint_id": dp.id,
+                                "run_id": run.id,
+                                "log_status": "complete",
+                            }
+                        )
+                        evaluation_context._callback(log.id)
+                except HumanloopRuntimeError as e:
+                    raise e
+                except Exception as e:
+                    log = log_func(
+                        **{
+                            "inputs": dp.inputs,
+                            "error": str(e),
+                            "source_datapoint_id": dp.id,
+                            "run_id": run.id,
+                            "start_time": start_time,
+                            "end_time": datetime.now(),
+                            "log_status": "complete",
+                        }
+                    )
+                    evaluation_context._callback(log.id)
+                    error_message = _get_error_message(e, length_limit=True)
+                    sys.stderr.write(
+                        f"\n{RED}Evaluated callable failed for Datapoint `{dp.id}`:\n{error_message}{RESET}\n"
+                    )
 
+        futures: list[Future] = []
         with ThreadPoolExecutor(max_workers=workers) as executor:
             for datapoint in hl_dataset.datapoints:
-                executor.submit(_process_datapoint, datapoint)
+                futures.append(executor.submit(_process_datapoint, datapoint))
+
+        for future in futures:
+            try:
+                future.result()
+            except Exception as e:
+                sys.stderr.write(f"\n{RED}Error processing datapoint:\n{_get_error_message(e)}{RESET}\n")
+                _cancel_evaluation()
```

Comment:
is it correct to cancel the whole eval when a single datapoint fails? seems excessive if I'm reading this right